client/src/session_manager.rs

```diff
@@ -569,23 +606,65 @@ impl SessionManager {
         node_id: &[u8],
         endpoints: &[proto::Endpoint],
         slot: u32,
+    ) -> anyhow::Result<NodeEntry> {
+        let node_id: NodeId = node_id.try_into()?;
+        let addrs: Vec<SocketAddr> = endpoints
+            .iter()
+            .cloned()
+            .filter_map(|e| e.try_into().ok())
+            .collect();
+
+        let target = self.guarded.get_or_create(node_id, &addrs).await;
+        let lock = target.lock.clone();
+        let _guard = lock.write().await;
+
+        // Check whether we are already connected to a node with this id
+        if let Ok(entry) = self.registry.resolve_node(node_id).await {
+            return Ok(entry);
+        }
+
+        // Check for own addresses
+        self.assert_remote_node(node_id, &addrs)
+            // Check whether we are already connected to a node with any of the addresses
+            .and_then(|_| self.assert_not_connected(&addrs))
+            .await?;
+
+        let this = self.clone();
+        self.try_resolve(node_id, addrs, slot)
+            .then(move |result| async move {
+                let mut guarded = this.guarded.state.write().await;
+                guarded.remove(&target);
```

Comment:
Since it is library, it depends on external code. Probably we don't drop anything at this moment, but I'm afraid what will happen after refactoring or future changes.
This problem is not easy to spot even when reading this function. It will be impossible to remember, in code that uses it

Reply:
`forward` is already prone to breaking the state when cancelled, the behavior is not unique to this change. The lock can still be acquired by consecutive or parallel calls

Reply:
You mean that there are no negative consequences?

Reply:
More data will be stored in memory in comparison to the current implementation. Other than that - the lock will be dropped, so the next call will be able acquire it and release the memory on completion